### DISCUSSION TITLE
Cannot create temporary directory

### DISCUSSION THREAD

**Discussion by lnfel** [norkunas/youtube-dl-php#176] [Q&A]
OS: Windows 10
Framework: Laravel 9
Using yt-dlp.exe standalone x64 binary version 2022.03.08.1

### I am receiving the following error:
`[err]: [12944] INTERNAL ERROR: cannot create temporary directory!`

I tried testing yt-dlp using command prompt on the same downloadPath and the video gets downloaded but for some reason I cannot run it using the youtube-dl-php wrapper. Code runs until the end and outputs an empty elements array of VideoCollection instance.

Ran `php --info` and saw `sys_temp_dir` has no value, dunno if this is the reason but it defaults to system %tmp% directory according to docs...

Also unchecked read-only in my downloads folder properties but it does not help at all.

### My code:
```php
use YoutubeDl\Options;
use YoutubeDl\YoutubeDl;
use Symfony\Component\Process\ExecutableFinder;

$yt = new YoutubeDl();

$yt->debug(function($type, $buffer) {
    echo "[$type]: $buffer";
});

$ytdlp = (new ExecutableFinder())->find('yt-dlp');
if (!$ytdlp) {
    $yt->setBinPath('C:\path-to\yt-dlp.exe');
}

$options = Options::create()
    ->downloadPath('C:\Downloads')
    ->url('https://www.youtube.com/watch?v=EngW7tLk6R8');

$collection = $yt->download($options);

foreach ($collection->getVideos() as $video) {
    if ($video->getError() !== null) {
        echo "Error downloading video: {$video->getError()}.";
    } else {
        echo $video->getTitle();
    }
}
```

Hoping to find a lead on this, in case somebody also experienced the same error.

Update:
Found this [thread](https://github.com/docker/compose/issues/3262) with the same error and they are talking about having low space on disk. Whipped up my cmder and check the disk with the following result:

```cmd
Size | Used | Avail | Used %
465G | 443G | 22G   | 96%
```

Kinda sus to me since 22G is still pretty big and I'm not using any VM.

**Comment by norkunas** [comment]
Hey.

It is strange, because this library does not maintain the download directory creation, but just delegates it to the yt-dlp/python.

Could you try to add a custom process builder and passing your downloads path to the process cwd argument?

```php
$yt = new YoutubeDl(new class implements ProcessBuilderInterface {
  public function build(?string $binPath, ?string $pythonPath, array $arguments = []): Process
  {
    array_unshift($arguments, $binPath);

    if ($pythonPath !== null) {
       array_unshift($arguments, $pythonPath);
    }

    $process = new Process(command: $arguments, cwd: 'C:\Downloads');
    $process->setTimeout(null);

    return $process;
  }
});
```

**Reply by lnfel** [reply]
Hi,

Thank you for the response.
I tried the custom process builder as you mentioned and the error is still happening. Tried to dump an instance of `Symfony\Component\Process\Process`, here is what it showed:

*`C:\Downloads` in this case is my project's public folder `C:\Sites\project-name\public\storage` which is equivalent to `base_path("public\storage")` in laravel

```php
Symfony\Component\Process\Process {
  -callback: null
  -hasCallback: false
  -commandline: array:7 [[▼]()
    0 => "C:\path-to\yt-dlp.EXE"
    1 => "--ignore-config"
    2 => "--ignore-errors"
    3 => "--write-info-json"
    4 => "--output=C:\Sites\project-name\public\storage/%(title)s-%(id)s.%(ext)s"
    5 => "https://www.youtube.com/watch?v=EngW7tLk6R8"
  ]
  -cwd: "C:\Sites\project-name\public\storage"
  -env: null
  -input: null
  -starttime: null
  -lastOutputTime: null
  -timeout: 60.0
  -idleTimeout: null
  -exitcode: null
  -fallbackStatus: []
  -processInformation: null
  -outputDisabled: false
  -stdout: null
  -stderr: null
  -process: null
  -status: "ready"
  -incrementalOutputOffset: 0
  -incrementalErrorOutputOffset: 0
  -tty: false
  -pty: false
  -options: array:2 [[▼]()
    "suppress_errors" => true
    "bypass_shell" => true
  ]
  -useFileHandles: true
  -processPipes: null
  -latestSignal: null
}
```

Here is the process builder:
```php
namespace App\Services\YoutubeDLP;

use Symfony\Component\Process\Process;
use YoutubeDl\Process\ProcessBuilderInterface;

class ProcessBuilder implements ProcessBuilderInterface
{
    public function build(?string $binPath, ?string $pythonPath, array $arguments = []):Process
    {
        array_unshift($arguments, $binPath);

        if ($pythonPath !== null) {
            array_unshift($arguments, $pythonPath);
        }

        $process = new Process(command: $arguments, cwd: base_path("public\storage"));
        dump($process);
        $process->setTimeout(null);

        return $process;
    }
}
```

And this is the updated code of my previous post:
```php
use App\Services\YoutubeDLP\ProcessBuilder;
use YoutubeDl\Options;
use YoutubeDl\YoutubeDl;
use Symfony\Component\Process\ExecutableFinder;

$yt = new YoutubeDl(new ProcessBuilder());

$yt->debug(function($type, $buffer) {
    echo "[$type]: $buffer";
});

$ytdlp = (new ExecutableFinder())->find('yt-dlp');

if ($ytdlp) {
    $yt->setBinPath($ytdlp);
}

$options = Options::create()
    ->downloadPath(base_path("public\storage"))
    ->url('https://www.youtube.com/watch?v=EngW7tLk6R8');

$collection = $yt->download($options);

foreach ($collection->getVideos() as $video) {
    ...omitted...
}
```

Still can't figure out why this is happening, I will try to dig more info if I can and post it here.

**Reply by norkunas** [reply]
If you'll find the solution to this problem, please share your findings, it may be useful for others :)
As I am not using Windows, I can't help you here :disappointed:
